### ISSUE TITLE
Not compatible with Angular Standalone Components

### ISSUE THREAD

**Issue by shaun-jacks** [storybookjs/testing-angular#2]
**Describe the bug**
When using this library for angular standalone components and storybook v7 (needed for the storybook 7), I get an error stating that my standalone component should not be in the declarations array of an NgModule.

This NgModule is the same module that is created to mount the component
**To Reproduce**
1. Using this library with Angular standalone components

**Expected behavior**
1. The library should create the mountable NgModule with a standalone component in the imports array if it is standalone, and declarations array if not

**Additional context**

**Comment by Marklb** [comment]
I haven't been great at maintaining this repo, but I am going to try and get it updated for the Storybook 7 changes soon. I have been thinking about it, while I was finishing up a change I have been working on for `@storybook/angular` and I think I know what needs to be done here, but I need to test it. I have something I need to finish by Friday for work, so I can't guarantee I will get to this before the weekend, but it is on my list of things I need to get caught up on.

The renderer has changed to return a standalone component, instead of an NgModule, so there will most likely be a change in how the library is used, but I don't expect it to be a difficult migration unless something more than simply rendering the story component is being done. I am considering supporting both standalone and NgModule usage through this, if I can, since I don't want to limit how this library can be used by testing framework renderers that may be used, but  I can't guarantee that at the moment.

**Comment by shaun-jacks** [comment]
That does makes sense, so it's potentially possible to use the library to render a standalone component instead of an NgModule. 

And I understand, thank you for the update and quick response!

**Comment by matmeylan** [comment]
It seems it's not just standalone components, but we also see issues non standalone components and storybook 7.

**Comment by shaun-jacks** [comment]
Any updates on a fix for this issue? 

I've investigated other workarounds such as using storybooks test runner, but it has come at the cost of losing Jest features and tooling.

**Comment by Marklb** [comment]
I may need to make some adjustments to Storybook to implement it the way I would like, but I have been having some issues with the Storybook repo that have been delaying that. What I have right now, should mostly work with v7, but I still need to finish testing more scenarios and updating examples.

~~I will try to get it published under the storybook namespace, within the next few days, but until I get the release updated, I published it to `@marklb/storybook-testing-angular@1.0.0-next.0`.~~ It is published under the storybook namespace now. `@storybook/testing-angular@1.0.1-next.1`

That version has only been tested on very simple tests, with Karma. I don't see a reason that it wouldn't work in Jest, but I am going to test that next. *I just tried it in one of my projects that uses Jest and it worked, but I only updated simple tests, so far.*

Currently, it only returns a standalone component and applicationConfig, but the `applicationConfig.providers` are already provided in the standalone component. I still need to test more complicated scenarios, to make sure that isn't limiting anything.

Because of changes to Storybook v7, I don't think the component is currently exposed where I need it, with how this works. I am working on a fix that adds it to a place I can access in the composed story, but this version needs it passed as the second parameter to `createMountableStoryComponent`.

Simple example with my current v7 changes:
```ts
import { render, screen } from '@testing-library/angular';
import { composeStory, createMountableStoryComponent } from '@storybook/testing-angular';

import * as stories from './button.component.stories';
import Meta from './button.component.stories';

const Primary = composeStory<ButtonComponent>(stories.Primary, Meta);
const Other = composeStory<ButtonComponent>(stories.Other, Meta);

describe('button testing-library', () => {
  it('renders primary button', async () => {
    const { component } = createMountableStoryComponent(Primary({}, {} as any), Meta.component);
    await render(component);
    expect(screen.getByText(Primary.args?.label!)).not.toBeNull();
  });
});
```

**Comment by shaun-jacks** [comment]
Really appreciate the update and the current API you have implemented at this time! I can try this out as well and let you know if I run into any issues.  

Just wanted to double check, in your example, is `Meta` the object that is defining the `applicationConfig.providers` that go with the standalone component?

**Comment by Marklb** [comment]
@shaun-jacks `Meta` is the default export in a stories CSF file. I probably shouldn't have used the same name as it's type. It is where you typically define the `title`, `component` and anything else that is applied to every story in the file.

**Comment by Marklb** [comment]
I published the current build under the storybook namespace now `@storybook/testing-angular@1.0.1-next.1`. I am going to try to fix the composed stories to include the component, so it doesn't have to be passed to `createMountableStoryComponent`, and add more scenarios to the examples this is tested on, then I will publish it to `latest`.

**Comment by shaun-jacks** [comment]
Just wanted to add a test scenario and error I ran into testing this out!

I just installed `@storybook/testing-angular@1.0.1-next.1`, the issue happens when my storybook needs  to import an EnvironmentProvider, in this example, an Ngrx Store provided in the component. When I inject the Store in the Meta decorators storybook, I get the error:

"NG0207: Invalid providers present in a non-environment injector. 'EnvironmentProviders' can't be used for component providers."

Here's an example:

```typescript
// card.component.stories.ts
import { Meta, applicationConfig } from '@storybook/angular';
import { CardComponent } from './card.component';
import { provideStore } from '@ngrx/store';

export default {
    title: 'CardComponent',
    component: CardComponent,
    decorators: [
        applicationConfig({
            providers: [provideStore({})],
        }),
    ],
} as Meta<CardComponent>;

export const Primary = {
    render: (args: CardComponent) => ({
        props: args,
    }),
    args: {},
};
```

```typescript
// card.component.spec.ts
import Meta from './card.component.stories';
import { composeStory, createMountableStoryComponent } from '@storybook/testing-angular';
import { render } from '@testing-library/angular';
import * as stories from './card.component.stories';
import { CardComponent } from './card.component';
const Primary = composeStory<CardComponent>(stories.Primary as any, Meta);

describe('interactive stories test', () =>
    it('should render and validate story', async () => {
        const { component } = createMountableStoryComponent(Primary({}, {}), Meta.component);
        await render(component);
    }));
```

I am currently using
 
- jest@29.4.3
- @testing-library/angular@13.4.0
- @storybook/testing-angular@1.0.1-next.1
- storybook@7.0.18
- @angular/core@16.0.2

**Comment by Marklb** [comment]
@shaun-jacks That is one of the scenarios I was worried about, when I tried to avoid the need to add the application providers yourself. A wrapper render method could probably be written to avoid that boilerplate, but I may just add something about doing that to the README, since it would depend on which testing framework being used.

I just published `@storybook/testing-angular@1.0.1-next.2`, which would now change your test to:
```ts
describe('interactive stories test', () =>
    it('should render and validate story', async () => {
        const { component, applicationConfig } = createMountableStoryComponent(Primary({}, {}), Meta.component);
        await render(component, { providers: applicationConfig.providers });
    }));
```

**Comment by shaun-jacks** [comment]
Appreciate the information, I see,  I just tried using the API you mentioned 

```typescript
const { component, applicationConfig } = createMountableStoryComponent(Primary({}, {}), Meta.component);
await render(component, { providers: applicationConfig.providers });
```

and it's working for my standalone components!

**Comment by shaun-jacks** [comment]
Would we be able to publish a version that is not a pre-release or next tagged version?

**Comment by Marklb** [comment]
@shaun-jacks Sorry for not already doing that. ~~I just published `@storybook/testing-angular@1.0.1`.~~

I just published `@storybook/testing-angular@1.0.2`. It has a workaround that should avoid the need for the redundant component passed to the second parameter of `createMountableStoryComponent`, so I made the second argument optional. I also added `createMountable`, which is shorter and excludes the redundant second parameter.
